test/libsolidity/astPropertyTests/forLoop/isSimpleCounterLoop.sol

```diff
@@ -0,0 +1,180 @@
+        }
+        /// FreeFunctionConditionRHS: isSimpleCounterLoop
+        for(uint i = 0; i < h(); ++i) {
+        }
+        // Negative Cases
+        /// AdditionLoopExpression: isSimpleCounterLoop
+        for(uint i = 0; i < 42; i = i + 1) {
+        }
+        /// SimplePreDecrement: isSimpleCounterLoop
+        for(uint i = 42; i > 0; --i) {
+        }
+        /// SimplePosDecrement: isSimpleCounterLoop
+        for(uint i = 42; i > 0; i--) {
+        }
+        /// MultiplicationLoopExpression: isSimpleCounterLoop
+        for(uint i = 1; i < 42; i = i * 2) {
+        }
+        /// CounterIncrementLoopBody: isSimpleCounterLoop
+        for(uint i = 0; i < 42; ++i) {
+            i++;
+        }
+        /// CounterAssignmentLoopBody: isSimpleCounterLoop
+        for(uint i = 0; i < 42; ++i) {
+            i = 43;
+        }
+        /// CounterAssignmentInlineAssemblyLoopBody: isSimpleCounterLoop
+        for(uint i = 0; i < 42; ++i) {
+            assembly {
+                i := add(i, 1)
+            }
+        }
+        uint j = type(uint).max;
+        /// ExternalCounterLoopExpression: isSimpleCounterLoop
+        for (uint i = 0; i < 10; ++j) {
+        }
+        /// CounterIncrementRHSAssignment: isSimpleCounterLoop
+        for(uint i = 0; i < 10; ++i) {
+            x = i++;
+        }
+        /// NoEffectLoopExpression: isSimpleCounterLoop
+        for(uint i = 0; i < 42; i) {
+        }
+        /// EmptyLoopExpression: isSimpleCounterLoop
+        for(uint i = 0; i < 10; ) {
+        }
+        uint y = type(uint8).max + 1;
+        /// DifferentCommonTypeCondition: isSimpleCounterLoop
+        for(uint8 i = 0; i < y; ++i) {
+        }
+        /// LessThanOrEqualCondition: isSimpleCounterLoop
+        for(uint i = 0; i <= 10; ++i) {
+        }
+        /// ComplexExpressionCondition: isSimpleCounterLoop
+        for(uint i = 0; (i < 10 || g()); ++i) {
+        }
+        /// FreeFunctionConditionLHS: isSimpleCounterLoop
+        for(uint i = 0; h() < 100; ++i) {
+        }
+        /// FreeFunctionConditionDifferentCommonTypeLHS: isSimpleCounterLoop
+        for(uint8 i = 0; i < h(); ++i) {
+        }
+        /// NonIntegerTypeCondition: isSimpleCounterLoop
+        for(uint i = 0; address(this) < msg.sender; ++i) {
+        }
+        /// UDVTOperators: isSimpleCounterLoop
+        for(UINT i = UINT.wrap(0); i < UINT.wrap(10); i = i + UINT.wrap(1)) {
+        }
+        /// CounterAssignmentConditionRHS: isSimpleCounterLoop
+        for(uint i = 0; i < (i = i + 1); ++i) {
+        }
+        /// LiteralDifferentCommonTypeConditionRHS: isSimpleCounterLoop
+        for(uint8 i = 0; i < 257 ; ++i) {
+        }
+        /// StateVarCounterModifiedFunctionConditionRHS: isSimpleCounterLoop
+        for (z = 1; z < modifyStateVarZ(); ++z) {
+        }
+        /// StateVarCounterModifiedFunctionLoopBody: isSimpleCounterLoop
+        for (z = 1; z < 2048; ++z) {
+            modifyStateVarZ();
```

Comment:
I think `LValueChecker` needs to check if invoked functions are pure/view only.

Reply:
Oh wow, yeah - I hadn't thought about state variables as loop variables actually :-). That won't happen in practice, but we indeed have to handle it properly nonetheless. But we can do something simpler: we can just require the loop counter to be a local variable (i.e. not a state variable) - that should be quick and painless to check and we won't loose any real-world cases.